File: compiler/nodes/statements/type_definition.go
package statements

import (
	"fmt"

	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/attributes"
	"wipple/nodes/types"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type DuplicateFieldDefinitionFact struct{}

func (fact DuplicateFieldDefinitionFact) String() string {
	return "is duplicate field definition"
}

type DuplicateVariantDefinitionFact struct{}

func (fact DuplicateVariantDefinitionFact) String() string {
	return "is duplicate variant definition"
}

type TypeDefinitionNode struct {
	Comments       []string
	Attributes     []*attributes.AttributeNode
	Name           string
	Parameters     []database.Node
	Representation TypeRepresentation
	Facts          *database.Facts
}

func (node *TypeDefinitionNode) GetFacts() *database.Facts {
	return node.Facts
}

type TypeAttributes struct {
	Intrinsic bool
}

type TypeRepresentation interface {
	GetSpan() database.Span
}

type StructureTypeRepresentation struct {
	Fields []FieldDefinition
	Span   database.Span
}

func (representation *StructureTypeRepresentation) GetSpan() database.Span {
	return representation.Span
}

type FieldDefinition struct {
	Name string
	Type database.Node
	Span database.Span
}

type EnumerationTypeRepresentation struct {
	Variants []VariantDefinition
	Span     database.Span
}

func (representation *EnumerationTypeRepresentation) GetSpan() database.Span {
	return representation.Span
}

type VariantDefinition struct {
	Name     string
	Elements []database.Node
	Span     database.Span
}

type MarkerTypeRepresentation struct {
	Span database.Span
}

func (representation *MarkerTypeRepresentation) GetSpan() database.Span {
	return representation.Span
}

func ParseTypeDefinitionStatement(parser *syntax.Parser) (*TypeDefinitionNode, *syntax.Error) {
	span := parser.Spanned()

	comments, err := ParseComments(parser)
	if err != nil {
		return nil, err
	}

	attributes, err := attributes.ParseAttributes(parser)
	if err != nil {
		return nil, err
	}

	name, err := syntax.ParseTypeName(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("AssignOperator")
	if err != nil {
		return nil, err
	}

	parser.ConsumeLineBreaks()

	parameters, err := types.ParseTypeParameters(parser)
	if err != nil {
		return nil, err
	}

	representation, err := ParseTypeRepresentation(parser)
	if err != nil {
		return nil, err
	}

	return &TypeDefinitionNode{
		Comments:       comments,
		Attributes:     attributes,
		Name:           name,
		Parameters:     typeParameterNodes(parameters),
		Representation: representation,
		Facts:          database.NewFacts(span()),
	}, nil
}

func ParseTypeAttributes(attrs []*attributes.AttributeNode) TypeAttributes {
	return TypeAttributes{
		Intrinsic: attributes.HasNameAttribute(attrs, "intrinsic"),
	}
}

func ParseTypeRepresentation(parser *syntax.Parser) (TypeRepresentation, *syntax.Error) {
	return syntax.ParseCached(parser, func(p *syntax.Parser) (TypeRepresentation, *syntax.Error) {
		structure, ok, err := syntax.ParseOptional(parser, ParseStructureTypeRepresentation)
		if err != nil {
			return nil, err
		}
		if ok {
			return structure, nil
		}

		enumeration, ok, err := syntax.ParseOptional(parser, ParseEnumerationTypeRepresentation)
		if err != nil {
			return nil, err
		}
		if ok {
			return enumeration, nil
		}

		marker, ok, err := syntax.ParseOptional(parser, ParseMarkerTypeRepresentation)
		if err != nil {
			return nil, err
		}
		if ok {
			return marker, nil
		}

		return nil, parser.Error("Expected type representation")
	})
}

func ParseStructureTypeRepresentation(parser *syntax.Parser) (*StructureTypeRepresentation, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("TypeKeyword")
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("LeftBrace")
	if err != nil {
		return nil, err
	}

	fields, err := syntax.ParseLines(parser, 1, true, ParseFieldDefinition)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightBrace")
	if err != nil {
		return nil, err
	}

	return &StructureTypeRepresentation{
		Fields: fields,
		Span:   span(),
	}, nil
}

func ParseFieldDefinition(parser *syntax.Parser) (FieldDefinition, *syntax.Error) {
	span := parser.Spanned()

	name, err := syntax.ParseVariableName(parser)
	if err != nil {
		return FieldDefinition{}, err
	}

	_, err = parser.Token("AnnotateOperator", syntax.TokenConfig{
		Commit: "in this type annotation",
	})
	if err != nil {
		return FieldDefinition{}, err
	}

	parser.ConsumeLineBreaks()

	fieldType, err := types.ParseType(parser)
	if err != nil {
		return FieldDefinition{}, err
	}

	return FieldDefinition{
		Name: name,
		Type: fieldType,
		Span: span(),
	}, nil
}

func ParseEnumerationTypeRepresentation(parser *syntax.Parser) (*EnumerationTypeRepresentation, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("TypeKeyword")
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("LeftBrace")
	if err != nil {
		return nil, err
	}

	variants, err := syntax.ParseLines(parser, 1, true, ParseVariantDefinition)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightBrace")
	if err != nil {
		return nil, err
	}

	return &EnumerationTypeRepresentation{
		Variants: variants,
		Span:     span(),
	}, nil
}

func ParseMarkerTypeRepresentation(parser *syntax.Parser) (*MarkerTypeRepresentation, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("TypeKeyword", syntax.TokenConfig{
		Commit: "in this type definition",
	})
	if err != nil {
		return nil, err
	}

	return &MarkerTypeRepresentation{
		Span: span(),
	}, nil
}

func ParseVariantDefinition(parser *syntax.Parser) (VariantDefinition, *syntax.Error) {
	span := parser.Spanned()

	name, err := syntax.ParseConstructorName(parser)
	if err != nil {
		return VariantDefinition{}, err
	}

	many, ok, err := syntax.ParseOptional(parser, func(parser *syntax.Parser) ([]syntax.Many[database.Node, struct{}], *syntax.Error) {
		return syntax.ParseMany(parser, 1, types.ParseAtomicType, syntax.ParseNothing)
	})
	if err != nil {
		return VariantDefinition{}, err
	}

	var elements []database.Node
	if ok {
		elements = make([]database.Node, 0, len(many))
		for _, item := range many {
			elements = append(elements, item.Value)
		}
	}

	return VariantDefinition{
		Name:     name,
		Elements: elements,
		Span:     span(),
	}, nil
}

func typeParameterNodes(parameters []types.TypeParameterNode) []database.Node {
	nodes := make([]database.Node, len(parameters))
	for i := range parameters {
		nodes[i] = &parameters[i]
	}

	return nodes
}

func (node *TypeDefinitionNode) Visit(visitor *visit.Visitor) {
	visit.Defining(visitor, node, func() (*visit.TypeDefinition, bool) {
		visitor.PushScope()

		visitor.CurrentDefinition.WithImplicitTypeParameters(func() {
			for _, parameter := range node.Parameters {
				visitor.Visit(parameter)
			}
		})

		visitor.Constraint(typecheck.NewTypeConstraint(node, typecheck.NamedType(node, node.Name, node.Parameters)))

		typeConstraints := visit.GetDefinitionConstraints(node)

		// Types don't have additional constraints

		definition := &visit.TypeDefinition{
			Name:       node.Name,
			Node:       node,
			Comments:   node.Comments,
			Attributes: visit.ParseTypeAttributes(node.Attributes),
			Parameters: node.Parameters,
		}

		if !definition.Attributes.Intrinsic {
			visitor.AfterTypeDefinitions(func() {
				switch representation := node.Representation.(type) {
				case *MarkerTypeRepresentation:
					visitor.PopScope()

					visitor.Define(node.Name, &visit.MarkerConstructorDefinition{
						Name:     definition.Name,
						Node:     definition.Node,
						Comments: definition.Comments,
					})
				case *StructureTypeRepresentation:
					fields := make(map[string]database.Node, len(representation.Fields))
					for _, field := range representation.Fields {
						visitor.Visit(field.Type)

						if _, ok := fields[field.Name]; ok {
							database.SetFact(node, DuplicateFieldDefinitionFact{})
							continue
						}

						fields[field.Name] = field.Type
					}

					visitor.PopScope()

					visitor.Define(node.Name, &visit.StructureConstructorDefinition{
						Name:     definition.Name,
						Node:     definition.Node,
						Comments: definition.Comments,
						Fields:   fields,
					})
				case *EnumerationTypeRepresentation:
					variantDefinitions := make(map[string]*visit.VariantConstructorDefinition, len(representation.Variants))
					for i, variant := range representation.Variants {
						if _, ok := variantDefinitions[variant.Name]; ok {
							database.SetFact(node, DuplicateVariantDefinitionFact{})
							continue
						}

						variantNode := &variantNode{
							Facts:   database.NewFacts(variant.Span),
							Index:   i,
							Variant: variant,
						}

						visitor.Db.Register(variantNode)

						database.SetFact(variantNode, typecheck.TypedFact{})

						constructorDefinition, _ := visit.Defining(visitor, variantNode, func() (*visit.VariantConstructorDefinition, bool) {
							// Inherit the type definition's constraints
							for _, constraint := range typeConstraints {
								visitor.Constraint(constraint)
							}

							for _, element := range variant.Elements {
								visitor.Visit(element)
							}

							if len(variant.Elements) == 0 {
								visitor.Constraint(typecheck.NewGroupConstraint(variantNode, node))
							} else {
								visitor.Constraint(typecheck.NewTypeConstraint(variantNode, typecheck.FunctionType[database.Node](variant.Elements, node)))
							}

							constructorDefinition := &visit.VariantConstructorDefinition{
								Name:  variant.Name,
								Node:  variantNode,
								Index: i,
							}

							return constructorDefinition, true
						})

						variantDefinitions[variant.Name] = constructorDefinition
					}

					visitor.PopScope()

					for name, constructorDefinition := range variantDefinitions {
						visitor.Define(name, constructorDefinition)
					}
				}
			})
		}

		visitor.PopScope()

		visitor.Define(node.Name, definition)

		return definition, true
	})
}

func (node *TypeDefinitionNode) Codegen(c *codegen.Codegen) error {
	// Handled specially in `codegen.Codegen`
	return nil
}

type variantNode struct {
	Facts   *database.Facts
	Index   int
	Variant VariantDefinition
}

func (node *variantNode) GetFacts() *database.Facts {
	return node.Facts
}

func (node *variantNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)

	elementTemporaries := make([]database.Node, len(node.Variant.Elements))
	for i, element := range node.Variant.Elements {
		temporary := &database.HiddenNode{
			Facts: database.NewFacts(database.GetSpanFact(element)),
		}

		c.Db.Register(temporary)

		elementTemporaries[i] = temporary
	}

	if len(elementTemporaries) == 0 {
		c.WriteString(span, fmt.Sprintf("__wipple_variant(%d, [])", node.Index))
	} else {
		c.WriteString(span, "(async (")
		for _, temporary := range elementTemporaries {
			c.WriteNode(span, temporary)
			c.WriteString(span, ", ")
		}
		c.WriteString(span, ") => {")
		c.WriteLine()

		c.WriteString(span, fmt.Sprintf("return __wipple_variant(%d, [", node.Index))
		for _, temporary := range elementTemporaries {
			c.WriteNode(span, temporary)
			c.WriteString(span, ", ")
		}
		c.WriteString(span, "]);")
		c.WriteLine()

		c.WriteString(span, "})")
	}

	return nil
}

File: compiler/src/nodes/statements/type_definition.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Db, Fact, HiddenNode, Node, NodeRef, Render, Span},
    nodes::{
        parse_atomic_type, parse_attributes, parse_comments, parse_type, parse_type_parameters,
    },
    syntax::{
        ParseError, Parser, TokenKind, parse_constructor_name, parse_type_name, parse_variable_name,
    },
    typecheck::{GroupConstraint, TypeConstraint, Typed},
    visit::{
        MarkerConstructorDefinition, StructureConstructorDefinition, TypeAttributes,
        TypeDefinition, VariantConstructorDefinition, Visit, Visitor,
    },
};
use std::collections::{BTreeMap, HashMap};

#[derive(Debug, Clone)]
pub struct DuplicateFieldDefinition;

impl Fact for DuplicateFieldDefinition {}

impl Render for DuplicateFieldDefinition {
    fn write(&self, w: &mut dyn std::fmt::Write, _db: &Db) -> std::fmt::Result {
        write!(w, "is duplicate field definition")
    }
}

#[derive(Debug, Clone)]
pub struct DuplicateVariantDefinition;

impl Fact for DuplicateVariantDefinition {}

impl Render for DuplicateVariantDefinition {
    fn write(&self, w: &mut dyn std::fmt::Write, _db: &Db) -> std::fmt::Result {
        write!(w, "is duplicate variant definition")
    }
}

#[derive(Debug, Clone)]
pub enum TypeRepresentation {
    Structure(StructureTypeRepresentation),
    Enumeration(EnumerationTypeRepresentation),
    Marker(MarkerTypeRepresentation),
}

#[derive(Debug, Clone)]
pub struct StructureTypeRepresentation {
    pub fields: Vec<FieldDefinition>,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct FieldDefinition {
    pub name: String,
    pub ty: NodeRef,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct EnumerationTypeRepresentation {
    pub variants: Vec<VariantDefinition>,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct VariantDefinition {
    pub name: String,
    pub elements: Vec<NodeRef>,
    pub span: Span,
}

#[derive(Debug, Clone)]
pub struct MarkerTypeRepresentation {
    pub span: Span,
}

#[derive(Debug)]
pub struct TypeDefinitionNode {
    pub comments: Vec<String>,
    pub attributes: Vec<NodeRef>,
    pub name: String,
    pub parameters: Vec<NodeRef>,
    pub representation: TypeRepresentation,
}

impl Node for TypeDefinitionNode {}

pub fn parse_type_definition_statement(
    parser: &mut Parser<'_>,
) -> Result<TypeDefinitionNode, ParseError> {
    let comments = parse_comments(parser)?;
    let attributes = parse_attributes(parser)?;
    let name = parse_type_name(parser)?;

    parser.token(TokenKind::AssignOperator)?;
    parser.consume_line_breaks();

    let parameters = parse_type_parameters(parser)?;
    let representation = parse_type_representation(parser)?;

    Ok(TypeDefinitionNode {
        comments,
        attributes,
        name,
        parameters,
        representation,
    })
}

pub fn parse_type_representation(
    parser: &mut Parser<'_>,
) -> Result<TypeRepresentation, ParseError> {
    parser.parse_cached(|parser| {
        if let Some(representation) = parser.parse_optional(parse_structure_type_representation)? {
            return Ok(TypeRepresentation::Structure(representation));
        }

        if let Some(representation) =
            parser.parse_optional(parse_enumeration_type_representation)?
        {
            return Ok(TypeRepresentation::Enumeration(representation));
        }

        if let Some(representation) = parser.parse_optional(parse_marker_type_representation)? {
            return Ok(TypeRepresentation::Marker(representation));
        }

        Err(parser.error("Expected type representation"))
    })
}

pub fn parse_structure_type_representation(
    parser: &mut Parser<'_>,
) -> Result<StructureTypeRepresentation, ParseError> {
    let span = parser.spanned();

    parser.token(TokenKind::TypeKeyword)?;
    parser.token(TokenKind::LeftBrace)?;

    let fields = parser.parse_lines(1, true, parse_field_definition)?;

    parser.token(TokenKind::RightBrace)?;

    Ok(StructureTypeRepresentation {
        fields,
        span: span(parser),
    })
}

pub fn parse_field_definition(parser: &mut Parser<'_>) -> Result<FieldDefinition, ParseError> {
    let span = parser.spanned();

    let name = parse_variable_name(parser)?;
    parser.token(TokenKind::AnnotateOperator)?;
    parser.commit("in this type annotation");
    parser.consume_line_breaks();
    let ty = parse_type(parser)?;

    Ok(FieldDefinition {
        name,
        ty,
        span: span(parser),
    })
}

pub fn parse_enumeration_type_representation(
    parser: &mut Parser<'_>,
) -> Result<EnumerationTypeRepresentation, ParseError> {
    let span = parser.spanned();

    parser.token(TokenKind::TypeKeyword)?;
    parser.token(TokenKind::LeftBrace)?;

    let variants = parser.parse_lines(1, true, parse_variant_definition)?;

    parser.token(TokenKind::RightBrace)?;

    Ok(EnumerationTypeRepresentation {
        variants,
        span: span(parser),
    })
}

pub fn parse_marker_type_representation(
    parser: &mut Parser<'_>,
) -> Result<MarkerTypeRepresentation, ParseError> {
    let span = parser.spanned();

    parser.token(TokenKind::TypeKeyword)?;
    parser.commit("in this type definition");

    Ok(MarkerTypeRepresentation { span: span(parser) })
}

pub fn parse_variant_definition(parser: &mut Parser<'_>) -> Result<VariantDefinition, ParseError> {
    let span = parser.spanned();

    let name = parse_constructor_name(parser)?;

    let elements = if let Some(elements) = parser.parse_optional(|parser| {
        parser.parse_many(1, parse_atomic_type, |parser| parser.parse_nothing())
    })? {
        elements.into_iter().map(|(node, _)| node).collect()
    } else {
        Vec::new()
    };

    Ok(VariantDefinition {
        name,
        elements,
        span: span(parser),
    })
}

impl Visit for TypeDefinitionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visitor.defining(node, |visitor| {
            visitor.push_scope();

            visitor.with_implicit_type_parameters(|visitor| {
                for parameter in &self.parameters {
                    visitor.visit(parameter);
                }
            });

            visitor.constraint(TypeConstraint::new(
                node.clone(),
                visitor.named_type(node.clone(), self.parameters.iter().cloned()),
            ));

            let type_constraints = visitor.current_definition().constraints(visitor.db);

            let definition = TypeDefinition {
                name: self.name.clone(),
                node: node.clone(),
                comments: self.comments.clone(),
                attributes: TypeAttributes::from_attributes(visitor.db, &self.attributes),
                parameters: self.parameters.clone(),
            };

            if !definition.attributes.intrinsic {
                visitor.after_type_definitions({
                    let representation = self.representation.clone();
                    let definition = definition.clone();
                    let node = node.clone();
                    move |visitor| match representation {
                        TypeRepresentation::Marker(_) => {
                            visitor.pop_scope();

                            visitor.define(
                                &definition.name,
                                MarkerConstructorDefinition {
                                    name: definition.name.clone(),
                                    node: definition.node.clone(),
                                    comments: definition.comments.clone(),
                                },
                            );
                        }
                        TypeRepresentation::Structure(representation) => {
                            let mut fields = BTreeMap::new();
                            for field in representation.fields {
                                visitor.visit(&field.ty);
                                visitor.edge(&field.ty, &node, "field");

                                if fields.contains_key(&field.name) {
                                    visitor.insert(&node, DuplicateFieldDefinition);
                                    continue;
                                }

                                fields.insert(field.name, field.ty);
                            }

                            visitor.pop_scope();

                            visitor.define(
                                &definition.name,
                                StructureConstructorDefinition {
                                    name: definition.name.clone(),
                                    node: definition.node.clone(),
                                    comments: definition.comments.clone(),
                                    fields,
                                },
                            );
                        }
                        TypeRepresentation::Enumeration(representation) => {
                            let mut variant_definitions = HashMap::new();
                            for (index, variant) in representation.variants.into_iter().enumerate()
                            {
                                if variant_definitions.contains_key(&variant.name) {
                                    visitor.insert(&node, DuplicateVariantDefinition);
                                    continue;
                                }

                                let variant_node = visitor.node(
                                    variant.span.clone(),
                                    VariantNode {
                                        index,
                                        variant: variant.clone(),
                                    },
                                );

                                visitor.insert(&variant_node, Typed::default());

                                let constructor_definition =
                                    visitor.defining(&variant_node.clone(), |visitor| {
                                        for constraint in &type_constraints {
                                            visitor.constraint(constraint.clone());
                                        }

                                        for element in &variant.elements {
                                            visitor.visit(element);
                                            visitor.edge(element, &variant_node, "element");
                                        }

                                        if variant.elements.is_empty() {
                                            visitor.constraint(GroupConstraint::new(
                                                variant_node.clone(),
                                                node.clone(),
                                            ));
                                        } else {
                                            visitor.constraint(TypeConstraint::new(
                                                variant_node.clone(),
                                                visitor.function_type(
                                                    variant.elements.iter().cloned(),
                                                    node.clone(),
                                                ),
                                            ));
                                        }

                                        VariantConstructorDefinition {
                                            name: variant.name.clone(),
                                            node: variant_node,
                                            index,
                                        }
                                    });

                                variant_definitions
                                    .insert(variant.name.clone(), constructor_definition);
                            }

                            visitor.pop_scope();

                            for (name, constructor_definition) in variant_definitions {
                                visitor.define(&name, constructor_definition);
                            }
                        }
                    }
                });
            }

            visitor.pop_scope();

            visitor.define(&definition.name, definition.clone());

            definition
        });
    }
}

impl Codegen for TypeDefinitionNode {
    fn codegen(&self, _codegen: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        // Handled specially in `Codegen`
        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct VariantNode {
    pub index: usize,
    pub variant: VariantDefinition,
}

impl Node for VariantNode {}

impl Visit for VariantNode {
    fn visit(&self, _node: &NodeRef, _visitor: &mut Visitor<'_>) {}
}

impl Codegen for VariantNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        let element_temporaries = self
            .variant
            .elements
            .iter()
            .map(|element| {
                let span = ctx.db.span(element);
                ctx.db.node(span, HiddenNode(None))
            })
            .collect::<Vec<_>>();

        if element_temporaries.is_empty() {
            ctx.write_string(format!("__wipple_variant({}, [])", self.index));
        } else {
            ctx.write_string("(async (");
            for temporary in &element_temporaries {
                ctx.write_node(temporary);
                ctx.write_string(", ");
            }
            ctx.write_string(") => {");
            ctx.write_line();

            ctx.write_string(format!("return __wipple_variant({}, [", self.index));
            for temporary in &element_temporaries {
                ctx.write_node(temporary);
                ctx.write_string(", ");
            }
            ctx.write_string("]);");
            ctx.write_line();

            ctx.write_string("})");
        }

        Ok(())
    }
}
